### PR TITLE
[WIP] Enables 2 factor option for login

### PR DESCRIPTION
**Pull request by thatarchguy**
Saving this merge request for after the release. Check it out though and critic it in the meantime. 

I didn't update the readme yet though in fears of merge conflicts. Once you give the go ahead that the core code is satisfactory then I'll do up the readme. There will probably be a merge conflict with the build-release.sh file if we release before this is merged. Don't worry about that. 

I introduced another js framework called bootboxjs. Basically I didn't want to keep making modals for confirmation dialogues. Check it out and tell me what you think. 

Signed-off-by: Kevin <kevin@stealsyour.pw>

### CODE REVIEW

File: migrations/versions/e3a72926f808_.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+"""empty message
```
Comment (PrahlM93) [inline migrations/versions/e3a72926f808_.py]: Add a comment here describing what the migration is for

File: migrations/versions/e3a72926f808_.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+# revision identifiers, used by Alembic.
+revision = 'e3a72926f808'
+down_revision = 'c3803ac9b7dd'
+
+from alembic import op
+import sqlalchemy as sa
+
+
+def upgrade():
+    ### commands auto generated by Alembic - please adjust! ###
```
Comment (PrahlM93) [inline migrations/versions/e3a72926f808_.py]: Remove this comment

File: migrations/versions/e3a72926f808_.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    op.add_column('postmaster_admins', sa.Column('otp_active', sa.Boolean(), nullable=True))
+    op.add_column('postmaster_admins', sa.Column('otp_secret', sa.String(length=16), nullable=True))
+    ### end Alembic commands ###
```
Comment (PrahlM93) [inline migrations/versions/e3a72926f808_.py]: Remove this comment

File: migrations/versions/e3a72926f808_.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+
+def downgrade():
+    ### commands auto generated by Alembic - please adjust! ###
```
Comment (PrahlM93) [inline migrations/versions/e3a72926f808_.py]: Remove this comment

File: migrations/versions/e3a72926f808_.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    op.drop_column('postmaster_admins', 'otp_secret')
+    op.drop_column('postmaster_admins', 'otp_active')
+    ### end Alembic commands ###
```
Comment (PrahlM93) [inline migrations/versions/e3a72926f808_.py]: Remove this comment

File: postmaster/apiv1/admins.py [hunk 6]
```diff
@@ -5,8 +5,10 @@
 an admin to create, delete, and update admins
 """
 
-from flask import request
+from flask import request, jsonify
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: jsonify should not be used because we have the json_wrap decorator which will jsonify dictionaries and SQLAlchemy queries

File: postmaster/apiv1/admins.py [hunk 7]
```diff
@@ -147,3 +149,112 @@ def unlock_admin(admin_id):
...
+@apiv1.route('/admins/<int:admin_id>/2factor', methods=['GET'])
+@login_required
+def twofactor_status(admin_id):
+    """ Returns if 2 factor is enabled or not
+
+        This information is in the main user route,
+        but I added it here as a stub for the URI.
+    """
+    admin = Admins.query.get_or_404(admin_id)
+    return jsonify({"enabled": admin.otp_active})
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: See comment on line 8

File: postmaster/apiv1/admins.py [hunk 8]
```diff
@@ -147,3 +149,112 @@ def unlock_admin(admin_id):
...
+                db.session.commit()
+            except ValidationError as e:
+                raise e
+            except Exception as e:
+                db.session.rollback()
+                json_logger(
+                    'error', current_user.username,
+                    'The following error occurred in twofactor_disable: {0}'.format(str(e)))
+                raise GenericError('The administrator could not be updated')
+            return jsonify({"enabled": admin.otp_active})
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: See comment on line 8

File: postmaster/apiv1/admins.py [hunk 9]
```diff
@@ -147,3 +149,112 @@ def unlock_admin(admin_id):
...
+        if admin.verify_totp(request.get_json(force=True).get('code')):
+            if not admin.otp_active:
+                auditMessage = 'The administrator "{0}" enabled 2 factor'.format(
+                    admin.username)
+                json_logger('audit', current_user.username, auditMessage)
+                admin.otp_active = True
+            try:
+                db.session.add(admin)
+                db.session.commit()
+                return jsonify({"status": "Success"})
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: See comment on line 8

File: postmaster/apiv1/admins.py [hunk 10]
```diff
@@ -147,3 +149,112 @@ def unlock_admin(admin_id):
     admin = Admins.query.get_or_404(admin_id)
     clear_lockout_fields_on_user(admin.username)
     return {}, 200
+
+
+@apiv1.route('/admins/<int:admin_id>/2factor', methods=['GET'])
+@login_required
+def twofactor_status(admin_id):
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: I'd rather not have this route and just have the user of the API query the admin itself.

File: postmaster/apiv1/admins.py [hunk 11]
```diff
@@ -147,3 +149,112 @@ def unlock_admin(admin_id):
...
+
+        Enabling 2 factor from this route is not possible.
+    """
+    admin = Admins.query.get_or_404(admin_id)
+    status = request.get_json(force=True).get('enabled')
+    if status:
+        if status.lower() == "false":
+            admin.otp_active = False
+            try:
+                db.session.add(admin)
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: This line isn't needed because admin is already associated with the session, but it is harmless to have it here.

File: postmaster/apiv1/admins.py [hunk 12]
```diff
@@ -147,3 +149,112 @@ def unlock_admin(admin_id):
...
+                raise e
+            except Exception as e:
+                db.session.rollback()
+                json_logger(
+                    'error', current_user.username,
+                    'The following error occurred in twofactor_disable: {0}'.format(str(e)))
+                raise GenericError('The administrator could not be updated')
+            return jsonify({"enabled": admin.otp_active})
+        elif status.lower() == "true":
+            raise GenericError("Cannot enable 2 factor from this route - see docs")
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: Should you potentially be able to toggle 2 factor on and off while retaining the last known OTP secret?

File: postmaster/apiv1/admins.py [hunk 13]
```diff
@@ -147,3 +149,112 @@ def unlock_admin(admin_id):
...
+    raise GenericError("Must provide 'enabled=False'")
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: Something more generic like:
An invalid parameter was supplied

File: postmaster/apiv1/admins.py [hunk 14]
```diff
@@ -147,3 +149,112 @@ def unlock_admin(admin_id):
...
+@login_required
+def qrcode(admin_id):
+    """ Presents the user with a QR code to scan to setup 2 factor authentication
+    """
+    # render qrcode for FreeTOTP
+    admin = Admins.query.get_or_404(admin_id)
+    if admin.id != current_user.id:
+        raise GenericError('You can only view your user\'s QR code')
+    if admin.otp_active:
+        return "Already gucci"
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: Get rid of this trashy language ;)

File: postmaster/apiv1/admins.py [hunk 15]
```diff
@@ -147,3 +149,112 @@ def unlock_admin(admin_id):
...
+    admin.generate_otp_secret()
+    try:
+        db.session.add(admin)
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: This line isn't needed because admin is already associated with the session, but it is harmless to have it here.

File: postmaster/models.py [hunk 16]
```diff
@@ -366,6 +371,17 @@ def set_password(self, new_password):
 
         self.password = bcrypt.generate_password_hash(new_password)
 
+    def generate_otp_secret(self, **kwargs):
+        # generate a random secret
+        self.otp_secret = base64.b32encode(urandom(10)).decode('utf-8')
+
+    def get_totp_uri(self):
```
Comment (PrahlM93) [inline postmaster/models.py]: This should probably throw an exception if self.otp_secret is not set

File: postmaster/static/js/admins.js [hunk 17]
```diff
@@ -190,6 +222,39 @@ function adminEventListeners () {
...
+      var id = $(e.relatedTarget).data('pk');
+      $.ajax({
+        url: '/api/v1/admins/' + id + '/2factor/qrcode',
+        success: function(data){
+           $("#qrcode").html(new XMLSerializer().serializeToString(data.documentElement));
+           verify2FABtn.attr('data-pk', id);
+        },
+        error: function (response) {
+          addStatusMessage('error', filterText(jQuery.parseJSON(response.responseText).message));
+          twoFAModal.modal('hide');
```
Comment (PrahlM93) [inline postmaster/static/js/admins.js]: This is very jarring where the modal is shown for a split second and then disappears with an error. The modal should only appear on success of loading the QR code

File: postmaster/templates/admins.html [hunk 18]
```diff
@@ -43,6 +45,31 @@ <h2 class="textHeadingManage">PostMaster Administrators</h2>
     <ul class="pagination hidden" id="pagination"></ul>
 </div>
 
+<!-- 2 Factor Modal -->
+<div class="modal fade" id="twoFAModal" role="dialog">
```
Comment (PrahlM93) [inline postmaster/templates/admins.html]: This should be converted to Bootbox since this was included as a dependency.

File: postmaster/templates/admins.html [hunk 19]
```diff
@@ -43,6 +45,31 @@ <h2 class="textHeadingManage">PostMaster Administrators</h2>
...
+            <div class="modal-footer">
+                <button type="button" class="btn btn-default" data-dismiss="modal">Cancel</button>
+
+            </div>
+        </div>
+    </div>
+</div>
+
 <!-- Confirmation Modal -->
 <div class="modal fade" id="deleteModal" role="dialog">
```
Comment (PrahlM93) [inline postmaster/templates/admins.html]: This should be converted to Bootbox since this was included as a dependency.

File: postmaster/templates/admins.html [hunk 20]
```diff
@@ -43,6 +45,31 @@ <h2 class="textHeadingManage">PostMaster Administrators</h2>
...
+<div class="modal fade" id="twoFAModal" role="dialog">
+    <div class="modal-dialog">
+        <div class="modal-content">
+            <div class="modal-header">
+                <h4 class="modal-title">Setup 2 Factor</h4>
+            </div>
+            <div class="modal-body">
+                <div id="qrcode"></div>
+                <p>Scan QR code with an App like Authy or Google Authenticator</p>
+                <div class="input-group">
```
Comment (PrahlM93) [inline postmaster/templates/admins.html]: This comes up looking goofy on my end:
image

File: postmaster/templates/admins.html [hunk 21]
```diff
@@ -43,6 +45,31 @@ <h2 class="textHeadingManage">PostMaster Administrators</h2>
...
 
+<!-- 2 Factor Modal -->
+<div class="modal fade" id="twoFAModal" role="dialog">
+    <div class="modal-dialog">
+        <div class="modal-content">
+            <div class="modal-header">
+                <h4 class="modal-title">Setup 2 Factor</h4>
+            </div>
+            <div class="modal-body">
+                <div id="qrcode"></div>
```
Comment (PrahlM93) [inline postmaster/templates/admins.html]: Let's center this

File: postmaster/apiv1/admins.py [hunk 22]
```diff
@@ -147,3 +149,112 @@ def unlock_admin(admin_id):
...
+            except Exception as e:
+                db.session.rollback()
+                json_logger(
+                    'error', current_user.username,
+                    'The following error occurred in verify_qrcode: {0}'.format(str(e)))
+                raise GenericError('The administrator could not be updated')
+        else:
+            raise GenericError("Invalid Code")
+    else:
+        raise ValidationError("Missing form parameter: code")
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: I'd like to reword this to:
The Input Code was not supplied

File: postmaster/apiv1/admins.py [hunk 23]
```diff
@@ -147,3 +149,112 @@ def unlock_admin(admin_id):
...
+            except ValidationError as e:
+                raise e
+            except Exception as e:
+                db.session.rollback()
+                json_logger(
+                    'error', current_user.username,
+                    'The following error occurred in verify_qrcode: {0}'.format(str(e)))
+                raise GenericError('The administrator could not be updated')
+        else:
+            raise GenericError("Invalid Code")
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: I'd like to reword this to:
An invalid code was supplied

File: postmaster/static/js/admins.js [hunk 24]
```diff
@@ -190,6 +222,39 @@ function adminEventListeners () {
...
+        method: 'POST',
+        contentType: 'application/json',
+        dataType: 'json',
+        data: JSON.stringify ({code: $("#verify2FACode").val()}),
+        success: function(data){
+          twoFAModal.modal('hide');
+          fillInTable();
+        },
+        error: function (response) {
+          addStatusMessage('error', filterText(jQuery.parseJSON(response.responseText).message));
```
Comment (PrahlM93) [inline postmaster/static/js/admins.js]: Either the modal should be closed on error, or the status message should be in the foreground (the messages are currently behind the modal)

File: postmaster/apiv1/admins.py [hunk 25]
```diff
@@ -147,3 +149,112 @@ def unlock_admin(admin_id):
...
+
+@apiv1.route('/admins/<int:admin_id>/2factor/qrcode', methods=['GET'])
+@login_required
+def qrcode(admin_id):
+    """ Presents the user with a QR code to scan to setup 2 factor authentication
+    """
+    # render qrcode for FreeTOTP
+    admin = Admins.query.get_or_404(admin_id)
+    if admin.id != current_user.id:
+        raise GenericError('You can only view your user\'s QR code')
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: Change this to:
You may only view your own QR code

File: postmaster/apiv1/admins.py [hunk 26]
```diff
@@ -147,3 +149,112 @@ def unlock_admin(admin_id):
...
+@login_required
+def verify_qrcode(admin_id):
+    """ Verifies if the 2 factor token provided is correct
+
+        This will enable 2 factor for a user.
+    """
+    admin = Admins.query.get_or_404(admin_id)
+    if request.get_json(force=True).get('code'):
+        if not admin.otp_secret:
+            raise GenericError("2 Factor Secret has not been generated yet")
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: Add "The" in front of this to conform to other messages.

File: postmaster/apiv1/admins.py [hunk 27]
```diff
@@ -147,3 +149,112 @@ def unlock_admin(admin_id):
...
+        if admin.verify_totp(request.get_json(force=True).get('code')):
+            if not admin.otp_active:
+                auditMessage = 'The administrator "{0}" enabled 2 factor'.format(
+                    admin.username)
+                json_logger('audit', current_user.username, auditMessage)
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: This should only be audited if the database commit was successful.

File: postmaster/apiv1/admins.py [hunk 28]
```diff
@@ -147,3 +149,112 @@ def unlock_admin(admin_id):
...
+
+        This will enable 2 factor for a user.
+    """
+    admin = Admins.query.get_or_404(admin_id)
+    if request.get_json(force=True).get('code'):
+        if not admin.otp_secret:
+            raise GenericError("2 Factor Secret has not been generated yet")
+        if admin.verify_totp(request.get_json(force=True).get('code')):
+            if not admin.otp_active:
+                auditMessage = 'The administrator "{0}" enabled 2 factor'.format(
```
Comment (PrahlM93) [inline postmaster/apiv1/admins.py]: All new Python code should have the variables in the style:
audit_message

File: postmaster/forms.py [hunk 29]
```diff
@@ -16,6 +16,7 @@ class LoginForm(Form):
     """
     username = StringField(label='Username', validators=[DataRequired()])
     password = PasswordField(label='Password', validators=[DataRequired(), validate_wtforms_password])
+    two_factor = IntegerField(label='2 Factor', validators=(validators.Optional(),))
```
Comment (PrahlM93) [inline postmaster/forms.py]: I'm not in love with this always being shown on the login screen as an input field. I would prefer if this was a modal that appeared after a successful password was entered.

We can discuss this further.

File: postmaster/static/js/admins.js [hunk 30]
```diff
@@ -279,6 +344,7 @@ function fillInTable () {
             html += '<td data-title="Username: "><a href="#" class="adminUsername" data-pk="' + item.id + '" data-url="/api/v1/admins/' + item.id + '" title="Click to change the username">' + filterText(item.username) + '</a></td>\
                     <td data-title="Password: "><a href="#" class="adminPassword" data-pk="' + item.id + '" data-url="/api/v1/admins/' + item.id + '" title="Click to change the password">●●●●●●●●</a></td>\
                     <td data-title="Name: "><a href="#" class="adminName" data-pk="' + item.id + '" data-url="/api/v1/admins/' + item.id + '" title="Click to change the name">' + filterText(item.name) + '</a></td>\
+                    <td data-title="2FA: ">' + (item.twoFactor ? ('<a href="#" class="admin2FADisable" data-pk="' + item.id + '" title="Click to disable 2-factor">Configured</a>') : ('<a href="#" class="admin2FAConfigure" data-pk="' + item.id + '" data-toggle="modal" data-target="#twoFAModal" title="Click to configure 2 Factor">Configure</a>')) + '</td>\
```
Comment (PrahlM93) [inline postmaster/static/js/admins.js]: I'd like to make this only show up in the row of the current admin. We can maybe add a field in the admin GET route that states whether or not this is the current admin that is logged in. That would give us the info we need to make the decision here.

File: postmaster/static/js/admins.js [hunk 31]
```diff
@@ -68,6 +87,7 @@ function editableAdminEventListeners() {
     var adminPassword = $('a.adminPassword');
     var adminName = $('a.adminName');
     var adminLocked = $('a.adminLocked')
+    var admin2FADisable = $('a.admin2FADisable')
```
Comment (PrahlM93) [inline postmaster/static/js/admins.js]: This should be a tooltip like the others

File: postmaster/static/js/admins.js [hunk 32]
```diff
@@ -44,21 +44,40 @@ function deleteAdmin(id) {
...
+// Disables 2FA for an administrator via the API
+function disable2FA(id, targetLink) {
+  $.ajax({
+    url: '/api/v1/admins/' + id + '/2factor',
+    method: 'PUT',
+    contentType: 'application/json',
+    dataType: 'json',
+    data: JSON.stringify({enabled: 'False'}),
+    success: function(data){
+      addStatusMessage('success', '2FA has been disabled');
```
Comment (PrahlM93) [inline postmaster/static/js/admins.js]: This should be something like:
Two factor authentication has been disabled

**Comment by PrahlM93** [comment]
@thatarchguy nice job on this. I have some comments/suggestions. Let me know if you want me to help or have any questions.

File: postmaster/apiv1/admins.py (hunk 6 above)
Comment (thatarchguy) [inline postmaster/apiv1/admins.py]: This doesn't work when returning a property of the object like I am.

File: postmaster/static/js/admins.js (hunk 17 above)
Comment (thatarchguy) [inline postmaster/static/js/admins.js]: Yeah I couldn't figure that one out. I'm not too good with jquery modals. I think the function shouldn't be on 'show.bs.modal'

File: postmaster/static/js/admins.js (hunk 30 above)
Comment (thatarchguy) [inline postmaster/static/js/admins.js]: That whole block of code is so dirty. I hate it.

Comment (PrahlM93) [inline postmaster/static/js/admins.js]: Yes I agree. I think we should turn this into a function that creates a table row in an object oriented manner.

File: postmaster/apiv1/admins.py (hunk 10 above)
Comment (thatarchguy) [inline postmaster/apiv1/admins.py]: Eh, I added in the comments why I made it. I think it's justified.

File: postmaster/apiv1/admins.py (hunk 12 above)
Comment (thatarchguy) [inline postmaster/apiv1/admins.py]: It shouldn't retain the last known OTP secret.

I didn't want to leave the user hanging if they were trying to hit it through this route. That is why I added that if statement to check if they were trying to enable it.

File: postmaster/models.py (hunk 16 above)
Comment (thatarchguy) [inline postmaster/models.py]: No reason to throw an exception, since the user will never be calling this function without generate_otp_secret being ran first. The exception would be for the developer, in which case we wouldn't want to catch it.

File: postmaster/templates/admins.html (hunk 20 above)
Comment (thatarchguy) [inline postmaster/templates/admins.html]: lgtm :dancer:

**Comment by mprahl** [comment]
@thatarchguy, I've been thinking about this, and maybe the thing to do here is split the pull request in two. One for the backend work, and one for the frontend work.

I think it's gotten to the point that we can't keep doing this dirty jQuery code and still have a UI that is manageable to maintain. Either we convert the UI to Angular 2, which I'm down (the only drawback is the lack of native x-editable), or we convert the table rows and columns to be jQuery objects.

**Comment by thatarchguy** [comment]
>We convert the table rows and columns to be jQuery objects
this seems much easier to do. I'm wondering if we should hold off converting to angular for another big release. 

The front-end stuff can be pull requested off of this pull request. Weird I know.

**Comment by mprahl** [comment]
What do you mean? haha
> The front-end stuff can be pull requested off of this pull request. Weird I know.

Okay, let's go ahead and do this with jQuery objects. I had started it a while back and it seemed like a much cleaner way of doing it than what we're doing now.

**Comment by thatarchguy** [comment]
You can branch off of this branch with the front-end stuff and create a pull request into this branch. Then this pull request would have all the changes.

**Comment by mprahl** [comment]
@thatarchguy I'm saying we should merge in the backend separately, then have a separate PR to master for the frontend.
